### ISSUE TITLE
Add icon name prop

### ISSUE THREAD

**Issue by ffpetrovic** [feathericons/react-feather#41]
Does it not make sense to have a reusable icon component of our own that utilizes react-feather, and therefore be able to do something like `<Icon icon={this.props.icon} />`?

I'm aware that we can always pass an icon like `<Camera />` down as a prop, but I'm wondering if it's possible to this just by the name of the icon.

**Comment by Kosai106** [comment]
You could always abstract this kind of behavour, but otherwise doing the following would allow you a similar experience (Almost).

```jsx
const IconTag = Icon[this.props.icon];

<IconTag />
```

**Comment by SeanMcP** [comment]
@Kosai106 I've created a component like that in every project where I've used `react-feather`. I agree with @ffpetrovic that baking that into the library would be a good idea.

**Comment by Kosai106** [comment]
I just realised that my method wouldn't work if you're using TypeScript anyway.

`Element implicitly has an 'any' type because type 'typeof import("/Users/oesterkilde/projects/[PROJECT-NAME]/node_modules/react-feather/dist/index")' has no index signature`

**Comment by LukyVj** [comment]
> @Kosai106 I've created a component like that in every project where I've used `react-feather`. I agree with @ffpetrovic that baking that into the library would be a good idea.

Thanks for the suggestion @Kosai106, this solved my issue :)

**Comment by LukyVj** [comment]
> You could always abstract this kind of behavour, but otherwise doing the following would allow you a similar experience (Almost).
> 
> ```js-jsx
> const IconTag = Icon[this.props.icon];
> 
> <IconTag />
> ```

Thanks for the suggestion @Kosai106, this solved my issue :)

**Comment by agustingabiola** [comment]
I know this is a bit old but in case anyone needs it I managed to create a component that lazy loads the icons based on the name (you'll need to use lowercase) and works with Typescript:

```import React, { Suspense } from 'react';

interface DpFeatherIconProps {
  icon: string;
}

export default (props: DpFeatherIconProps) => {
  const IconTag = React.lazy(() => import(`react-feather/dist/icons/${props.icon}.js`));

  return (
    <Suspense fallback={null}>
      <IconTag />
    </Suspense>
  );
};```

**Comment by kungpaogao** [comment]
@agustingabiola Thank you for this workaround. However, the use of lazy/Suspense seems to cause some "blinking" on some loads. Are there any other workarounds?

**Comment by Kosai106** [comment]
@kungpaogao For the `fallback`, just use an empty div with the same dimensions as your icon - This should prevent any layout shifting.

**Comment by kungpaogao** [comment]
@Kosai106 yup, that’s what I’m doing already, but it doesn’t solve the “blinking” that happens when the icon is loaded

**Comment by agustingabiola** [comment]
> @Kosai106 yup, that’s what I’m doing already, but it doesn’t solve the “blinking” that happens when the icon is loaded

Use memo for the component alongside with an element with the same dimensions for fallback. For example if you have a blue cross of 16px you can check that if all props are equal return the same rendered component

**Comment by kraenhansen** [comment]
For anyone interested, I just wrote this functional component which will render an icon by name (in TypeScript):

```typescript
import React from "react";
import icons, { Props } from "react-feather";

type IconName = keyof typeof icons;

export type IconProps = {
  name: IconName;
} & Props;

export function Icon({ name, ...rest }: IconProps) {
  const IconComponent = icons[name];
  return <IconComponent {...rest} />;
}
```

**Comment by kungpaogao** [comment]
@kraenhansen This is exactly what I'm looking for! Thank you!

**Comment by osvaldokalvaitir** [comment]
It would be very interesting to be able to pass on the name of the property

**Comment by LukyVj** [comment]
Hello :wave: 

Nice solution from @kraenhansen, however, using this on next@10 & TypeScript result with the following error: 

```
Server Error
Error: Element type is invalid: expected a string (for built-in components) 
or a class/function (for composite components) but got: undefined.
You likely forgot to export your component from the file it's defined in, 
or you might have mixed up default and named imports.
``` 


<details>
<img width="982" alt="Screenshot 2020-11-05 at 10 49 26" src="https://user-images.githubusercontent.com/1630200/98225167-af147700-1f54-11eb-960a-30a8efeea007.png">
</details>

If anyone got a clue or a solution for this. 

Below, my current files: 
<details>

`Icon/Icon.tsx`

```typescript
import React from "react";
import * as icons from "react-feather";

export type IconName = keyof typeof icons;

export type IconProps = {
  name: IconName;
} & icons.Props;

export function Icon({ name, ...rest }: IconProps) {
  const IconComponent = icons[name];
  return <IconComponent {...rest} />;
}
```

And where I use the icon: 
```typescript
import { Icon } from "../Icon/Icon";

...
...
...
{icon && <Icon name={icon} />}
```
</details>


From what I understand, the exports/imports seems to be wrong somehow, I've tried several ways to fix this but none works, any hint would be appreciated 😄

**Comment by Kosai106** [comment]
@LukyVj It works just fine in Next.js from what I can tell.
https://codesandbox.io/s/hopeful-franklin-xr1fe?file=/src/pages/index.tsx

From the error message you're sharing it seems to come from somewhere else.

**Comment by LukyVj** [comment]
> @LukyVj It works just fine in Next.js from what I can tell.
> [codesandbox.io/s/hopeful-franklin-xr1fe?file=/src/pages/index.tsx](https://codesandbox.io/s/hopeful-franklin-xr1fe?file=/src/pages/index.tsx)
> 
> From the error message you're sharing it seems to come from somewhere else.

Thanks for answering! 

So, yeah, I've tried your code, works well when I pass the name directly as a string `Icon name="box"` works just fine.

But, when I tried to pass it through a prop, it doesn't work, e.g:

Where I define the needed icon
```javascript
  <Button
        onClick={() => {
          toggleOpen(!open);
        }}
        primary
        className="mb-16"
        icon="Box"
      />
```

The button code: 
```
{icon && <Icon name={icon} />}
```

So, I've somehow understood why this doesn't work based on your comment, turns out my component get rendered several times including one where the `icon` is undefined, I suppose that's what's causing the error :D

**Comment by davidperklin** [comment]
> > @LukyVj It works just fine in Next.js from what I can tell.
> > [codesandbox.io/s/hopeful-franklin-xr1fe?file=/src/pages/index.tsx](https://codesandbox.io/s/hopeful-franklin-xr1fe?file=/src/pages/index.tsx)
> > From the error message you're sharing it seems to come from somewhere else.
> 
> Thanks for answering!
> 
> So, yeah, I've tried your code, works well when I pass the name directly as a string `Icon name="box"` works just fine.
> 
> But, when I tried to pass it through a prop, it doesn't work, e.g:
> 
> Where I define the needed icon
> 
> ```js
>   <Button
>         onClick={() => {
>           toggleOpen(!open);
>         }}
>         primary
>         className="mb-16"
>         icon="Box"
>       />
> ```
> 
> The button code:
> 
> ```
> {icon && <Icon name={icon} />}
> ```
> 
> So, I've somehow understood why this doesn't work based on your comment, turns out my component get rendered several times including one where the `icon` is undefined, I suppose that's what's causing the error :D

Have you figured out a solution for this?

**Comment by GCobo** [comment]
@kraenhansen thanks!!

**Comment by lawnchamp** [comment]
@kraenhansen doesn't that solution have the problem of all 280 icons being included into your webpack bundle so that you can pick components dynamically? Tree shaking wouldn't be able to work in this case right?

**Comment by BenSampo** [comment]
> For anyone interested, I just wrote this functional component which will render an icon by name (in TypeScript):
> 
> ```ts
> import React from "react";
> import * as icons from "react-feather";
> 
> export type IconName = keyof typeof icons;
> 
> export type IconProps = {
>   name: IconName;
> } & icons.Props;
> 
> export function Icon({ name, ...rest }: IconProps) {
>   const IconComponent = icons[name];
>   return <IconComponent {...rest} />;
> }
> ```

I have modified as follows, in case it's useful to anyone.

```ts
import React from "react";
import * as Icons from "react-feather";

export type IconName = keyof typeof Icons;

export type IconProps = {
    name: IconName;
} & Icons.IconProps;

const Icon: React.FC<IconProps> = ({ name, ...rest }) => {
    const IconComponent = Icons[name];
    return <IconComponent {...rest} />;
};

export default Icon;
```

Usage is `<Icon name="Download" />`
